examples/decoding/plot_ssd_spatial_filters.py

```diff
@@ -41,101 +50,346 @@ def freq_mask(freqs, fmin, fmax):
+        else:
+            self.n_fft=int(n_fft)
+        self.picks_ = picks
+        
+        self.return_filtered=return_filtered
+        
+       
         self.estimator = estimator
         self.n_components = n_components
-
+        self.rank = rank
+        self.sampling_freq=sampling_freq      
+        self.cov_method_params = cov_method_params
+        
+    
     def fit(self, inst):
         """Fit"""
-        if self.picks == None:
-            self.picks_ = mne.pick_types(inst.info, meg=True, eeg=False, ref_meg=False)
-        else:
-            self.picks_ = self.picks
-        if isinstance(inst, mne.io.base.BaseRaw):
+        
+        
+        if isinstance(inst, BaseRaw):
+            if self.picks_ is None:
+                raise ValueError('picks should be provided')
+            self.max_components=len(self.picks_)
             inst_signal = inst.copy()
             inst_signal.filter(picks=self.picks_, **self.filt_params_signal)
-
+            self.Xs=inst_signal
+            #noise
             inst_noise = inst.copy()
             inst_noise.filter(picks=self.picks_, **self.filt_params_noise)
-            if self.subtract_signal:
-                inst_noise._data[self.picks_] -= inst_signal._data[self.picks_]
-            cov_signal = mne.compute_raw_covariance(
-                inst_signal, picks=self.picks_, method=self.estimator)
-            cov_noise = mne.compute_raw_covariance(inst_noise, picks=self.picks_,
-                method=self.estimator)
+            # subtract signal:
+            inst_noise._data[self.picks_] -= inst_signal._data[self.picks_]
+           
+            cov_signal = compute_raw_covariance(
+                inst_signal, picks=self.picks_, method=self.estimator, rank=self.rank)
+            cov_noise = compute_raw_covariance(inst_noise, picks=self.picks_,
+                method=self.estimator, rank=self.rank)
             del inst_noise
             del inst_signal
         else:
-            raise NotImplementedError()
+            if isinstance(inst, BaseEpochs)  or isinstance(inst, np.ndarray):
+            # data X is epoched 
+            # part of the following code is copied from mne csp
+                n_epochs, n_channels, n_samples = inst.shape
+                self.max_components=n_channels
+              
+                #reshape for filtering
+                X_aux=np.reshape(inst, [n_epochs, n_channels*n_samples])
+                X_s=filter_data(X_aux, self.sampling_freq, **self.filt_params_signal)  
+                
+                #rephase for filtering
+                X_aux=np.reshape(inst, [n_epochs, n_channels*n_samples])
+                X_n=filter_data(X_aux, self.sampling_freq, **self.filt_params_noise)
+                # subtract signal:
+                X_n -= X_s
+                
+                # Estimate single trial covariance
+                #reshape to original shape
+                X_s=np.reshape(X_s, [n_epochs,n_channels,n_samples])
```

Comment:
This is another situation where whitespaces are needed: equals operator and commas. I would recommend installing style-checker plugins for the editor of choice.

Reply:
Great ! I have checked the code now with an style-checker!

Reply:
Sometimes they are missing something .. I used pyflakes and pep8. Here we had particularly many white space errors so I suspect something has not been working at that level.

Reply:
nono, I meant, I had now used for the updated version the style-checker (I didn't have it before activated) :)